### PR TITLE
feat(manager/pep621): support hatch environments

### PR DESCRIPTION
**Pull request by jonathanberthias**
## Changes

This PR extends the PEP621 manager to add support for updating dependencies specified in [`hatch` environments](https://hatch.pypa.io/latest/config/environment/overview/#dependencies).

This work builds on top of #22082 by adding a new processor. Note that `hatch` does not have lock files.

## Context

Closes #24956.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by secustor — COMMENTED** [review]
LGTM outside of the lint errors

File: lib/modules/manager/pep621/processors/hatch.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    return deps;
+  }
+
+  async updateArtifacts(
+    updateArtifact: UpdateArtifact<Record<string, unknown>>,
+    project: PyProject
+  ): Promise<UpdateArtifactsResult[] | null> {
+    // Hatch does not have lock files at the moment
+    // https://github.com/pypa/hatch/issues/749
+    return await null;
```
Comment (secustor) [inline lib/modules/manager/pep621/processors/hatch.ts:45]:
```suggestion
    return Promise.resolve(null);
```

**Review by secustor — CHANGES_REQUESTED** [review]
Run `pnpm test` to do a full acceptance test

**Comment by jonathanberthias** [comment]
> Run `pnpm test` to do a full acceptance test

For some reason it gets stuck forever at the `eslint` step 🤷

File: lib/modules/manager/pep621/extract.spec.ts [hunk 2]
```diff
@@ -4,6 +4,7 @@ import { extractPackageFile } from '.';
 
 const pdmPyProject = Fixtures.get('pyproject_with_pdm.toml');
 const pdmSourcesPyProject = Fixtures.get('pyproject_pdm_sources.toml');
+const hatchPyProject = Fixtures.get('pyproject_with_hatch.toml');
```
Comment (viceice) [inline lib/modules/manager/pep621/extract.spec.ts:7]: move to test, it's only used once

File: lib/modules/manager/pep621/processors/hatch.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  PackageDependency,
+  UpdateArtifact,
+  UpdateArtifactsResult,
+} from '../../types';
+import type { PyProject } from '../schema';
+import { parseDependencyList } from '../utils';
+import type { PyProjectProcessor } from './types';
+
+function depTypeFromEnv(envName: string): string {
+  return `tool.hatch.envs.${envName}`;
```
Comment (viceice) [inline lib/modules/manager/pep621/processors/hatch.ts:12]: inline this

File: lib/modules/manager/pep621/processors/hatch.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        env?.['extra-dependencies']
+      );
+      deps.push(...extraDeps);
+    }
+
+    return deps;
+  }
+
+  updateArtifacts(
+    updateArtifact: UpdateArtifact<Record<string, unknown>>,
```
Comment (viceice) [inline lib/modules/manager/pep621/processors/hatch.ts:40]:
```suggestion
    updateArtifact: UpdateArtifact,
```
it's the default, so no need to explicit define

**Review by secustor — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**